odoo/tools/float_utils.py

```diff
@@ -215,6 +215,23 @@ def float_split(value, precision_digits):
         return int(units), 0
     return int(units), int(cents)
 
+def format_float_round(value, precision_digits, rounding_method='HALF-UP'):
+    """ same as float_repr excepts that it returns a float representation suitable for json
+        this is needed because float_round operation will sometimes have a small error (i.e. 3.1750000000000003)
+        float(float_repr()) will work in most cases since standard library tries to keep eval(repr(f)) == f.
+        see: https://bugs.python.org/issue1580
+
+        :param int precision_digits: number of fractional digits to
+                                     include in the output.
+        :param rounding_method: the rounding method used: 'HALF-UP', 'UP' or 'DOWN',
+           the first one rounding up to the closest number with the rule that
+           number>=0.5 is rounded up to 1, the second always rounding up and the
+           latest one always rounding down.
+        :return: rounded float
+    """
+    return float(float_repr(float_round(value, precision_digits=precision_digits, rounding_method=rounding_method),
+                            precision_digits=precision_digits))
```

Comment:
:+1: Nice! Now, as naming things is very important for a new API method, I would suggest the following small changes:
- name it `json_float_round()` to make its usage super intuitive
- state clearly that it cannot be used for rounding during calculations, it should be the final step, and is to be used more like `float_repr`.
- document the rationale for doing this rather than having a custom Float class (it's because it doesn't work, json.dumps doesn't allow it)
- nice names for intermediary variables, to follow the style of `float_round`, and make the logic very easy to follow

```suggestion
def json_float_round(value, precision_digits, rounding_method='HALF-UP'):
    """Not suitable for float calculations! Similar to float_repr except that it
    returns a float suitable for json dump

    This may be necessary to produce "exact" representations of rounded float
    values during serialization, such as what is done by `json.dumps()`.
    Unfortunately `json.dumps` does not allow any form of custom float representation,
    nor any custom types, everything is serialized from the basic JSON types.

    :param int precision_digits: number of fractional digits to round to.
    :param rounding_method: the rounding method used: 'HALF-UP', 'UP' or 'DOWN',
           the first one rounding up to the closest number with the rule that
           number>=0.5 is rounded up to 1, the second always rounding up and the
           latest one always rounding down.
    :return: a rounded float value that must not be used for calculations, but
             is ready to be serialized in JSON with minimal chances of
             representation errors.
    """
    rounded_value = float_round(f, precision_digits=precision_digits, rounding_method=rounding_method)
    rounded_repr = float_repr(rounded_value, precision_digits=precision_digits)
    # As of Python 3.1, rounded_repr should be the shortest representation for our
    # rounded float, so we create a new float whose repr is expected
    # to be the same value, or a value that is semantically identical
    # and will be used in the json serialization.
    # e.g. if rounded_repr is '3.1750', the new float repr could be 3.175
    # but not 3.1740000000322452.
    # Cfr. bpo-1580: https://bugs.python.org/issue1580
    return float(rounded_repr) 
```